src/tree/tree.ts

```diff
@@ -118,32 +137,53 @@ export class SchemaTreeListTree extends TreeListTree {
     this.invalidate();
   }
 
-  protected createTreeNode(schemaNode: SchemaNode): TreeListNode {
+  protected createTreeNode(schemaNode: SchemaNode): SchemaTreeListNode {
     const parentTreeNode = schemaNode.parent === null ? null : this.getFromSchemaToTreeMap(schemaNode.parent);
     if (parentTreeNode !== null) {
       assertParentTreeNode(parentTreeNode);
     }
 
-    const treeNode: TreeListNode = {
+    const treeNode: SchemaTreeListNode = {
       id: schemaNode.id,
       parent: parentTreeNode,
       name: '',
-      metadata: schemaNode,
+      metadata: { schemaNode },
       ...('children' in schemaNode && isNonNullable(schemaNode.children) && { children: [] }),
     };
 
+    if (
+      isRegularNode(schemaNode) &&
+      ['anyOf', 'oneOf'].includes(schemaNode.combiners?.[0] ?? '') &&
+      schemaNode.children
+    ) {
+      const child = schemaNode.children[this._selectedIndex];
+      treeNode.metadata = {
+        schemaNode: child,
+        typeOptions: schemaNode.children,
+      };
+      if('children' in child && isNonNullable(child.children)){
+        (treeNode as SchemaTreeListNodeWithChildren).children = [];
+      }
+      this._schemaToTreeMap.set(child, treeNode);
+      this._treeToSchemaMap.set(treeNode, child);
+    } else {
+      this._schemaToTreeMap.set(schemaNode, treeNode);
+      this._treeToSchemaMap.set(treeNode, schemaNode);
+    }
+
     if (isParentTreeNode(treeNode.parent)) {
       this.visited.add(treeNode.parent);
     }
 
-    this._schemaToTreeMap.set(schemaNode, treeNode);
-    this._treeToSchemaMap.set(treeNode, schemaNode);
-
     return treeNode;
   }
 
-  public buildTreeFragment(schemaNode: SchemaNode) {
-    const treeNode = this.createTreeNode(schemaNode);
+  public buildTreeFragment(rawSchemaNode: SchemaNode): SchemaTreeListNode {
+    let treeNode: SchemaTreeListNode = this.createTreeNode(rawSchemaNode);
+    const schemaNode =
+      isRegularNode(rawSchemaNode) && ['anyOf', 'oneOf'].includes(rawSchemaNode.combiners?.[0] ?? '')
```

Comment:
Be careful.
`oneOf`/`anyOf` may undergo a merging process.
See https://github.com/stoplightio/json-schema-tree/blob/beta/src/walker/walker.ts#L287
I don't know how we want to tackle that behavior from a product perspective, but ti's certainly something to watch out for.
I suspect no changes are likely to be required in the piece of code I linked above, but you might want to verify that.